core/model/modx/transport/modtransportpackage.class.php

```diff
@@ -362,8 +362,26 @@ public function transferPackage($sourceFile, $targetDir) {
         if (is_dir($targetDir) && is_writable($targetDir)) {
             if (!is_array($this->xpdo->version)) { $this->xpdo->getVersionData(); }
             $productVersion = $this->xpdo->version['code_name'].'-'.$this->xpdo->version['full_version'];
-
-            $source= $this->get('service_url') . $sourceFile.(strpos($sourceFile,'?') !== false ? '&' : '?').'revolution_version='.$productVersion;
+            
+            /* make sure the package is downloaded, if not attempt re-download */
+            if (!file_exists($targetDir . $sourceFile)) {
+                /* not very pretty way of getting the download URL, but otherwise we have to make a 
+                 * remote request to $this->Provider->info($this->signature) to get the URL.
+                 * loop trough metadata arrays until location is found */
+                $locationarr = array_filter($this->get('metadata'), function($item) { 
+                    if ($item['name'] === 'location') { 
+                        return $item; 
+                    }
+                });
+                /* determine the index at which the location metadata was found */
+                $locationkey = array_keys($locationarr)[0];
```

Comment:
What happens if the array is empty after applying the filter?

Reply:
It would die... =)... question is, would this ever happen? Answer, probably...so maybe I use `reset()` to get the first element and check later if it's not empty...you right, should have taken care of that!